src/mongo/db/pipeline/expression_test.cpp

```diff
@@ -2247,6 +2247,45 @@ TEST(ExpressionArrayTest, ExpressionArrayShouldOptimizeSubExpressionToExpression
     ASSERT_TRUE(constantExpression);
 }
 
+TEST(ExpressionIndexOfArray, ExpressionIndexOfArrayShouldOptimizeArguments) {
+    intrusive_ptr<ExpressionContextForTest> expCtx(new ExpressionContextForTest());
+
+    auto expIndexOfArray = Expression::parseExpression(
+        expCtx,
+        BSON("$indexOfArray" << BSON_ARRAY(BSON_ARRAY(BSON("$add" << BSON_ARRAY(1 << 1)) << 1 << 1)
+                                           // Start index
+                                           << BSON("$add" << BSON_ARRAY(1 << 1))
+                                           // End index
+                                           << BSON("$add" << BSON_ARRAY(1 << 1)))),
+        expCtx->variablesParseState);
+    auto argsOptimizedToConstants = expIndexOfArray->optimize();
+    auto shouldBeNary = dynamic_cast<ExpressionNary*>(argsOptimizedToConstants.get());
+    ASSERT_TRUE(shouldBeNary);
+
+    auto optimizedArgs = shouldBeNary->getOperandList();
+    // All arguments should be ExpressionConstants.
```

Comment:
Oh, interesting - I would have expected it to fully optimize to a constant in this case... I wonder why it didn't? Can you make sure it does?

Reply:
I think this is because we are optimizing an ExpresssionIndexOfArray not evaluating so it should still be an ExpressionIndexOfArray.

Im casting it to an ExpressionNary because I need to be able to call getOperandList() to determine if the arguments optimized. I think this is where the confusion is to clearify I will cast it to an ExpressionIndexOfArray